### DISCUSSION TITLE
How do I measure text length with TextWriter? (`fitz.get_text_length` not recommended)

### DISCUSSION THREAD

**Discussion by cbm755** [pymupdf/PyMuPDF#1915] [Q&A] [answered]
I tried using the legacy `get_text_length` with TextWriter but this is a bad idea as `get_text_length` does not know about TextWriter's fancy fallback font support. 

For example:
```python
    tw = fitz.TextWriter(...)
    print(fitz.get_text_length("我爱PyMuPDF!", fontsize=36, fontname="Helv"))
    tw.append(fitz.Point(0, 100), "我爱PyMuPDF!", fontsize=36)
    print(tw.text_rect.width)
```
This gives 194.03999948501587 and 295.2773742675781, presumably b/c the first one would render as `??PyMuPDF!` using the old `insert_text` methods, whereas TextWriter has a fallback font.

My workaround is to instead construct a temporary `TextWriter` just to measure the text length, then discard it. (In my real problem, I want to adjust fontsize with a while loop if a string to too long to fit in a box).

Anyone see a better way?   And should the docs of `get_text_length` warn about this situation?

**Comment by JorjMcKie** [comment, marked as answer]
The correct (and fastest) way would be to use the PyMuPDF class `Font`. No need to employ TextWriter:
```python
import fitz
font = fitz.Font("helv")  # supports the same fontcodes as the traditional "insert_text"
# this this the correct number, and also uses the fallback feature:
font.text_length("我爱PyMuPDF!",fontsize=36)  
246.02399969100952
# crosscheck
font.text_length("PyMuPDF!",fontsize=36)
174.02399969100952
# difference:
font.text_length("我爱PyMuPDF!",fontsize=36)-font.text_length("PyMuPDF!",fontsize=36)
72.0
# which is plausible, because Chinese are monospace, 72 = 2 times fontsize

# crosscheck: use Droid Sans Fallback font:
font2=fitz.Font("cjk")
font2.text_length("PyMuPDF!",fontsize=36)
185.29200160503387
font2.text_length("我爱PyMuPDF!",fontsize=36)
257.2920016050339
font2.text_length("我爱PyMuPDF!",fontsize=36)-font2.text_length("PyMuPDF!",fontsize=36)
72.0
# if using text writer, use insertion END POINT to compare:
tw=fitz.TextWriter(fitz.paper_rect("a4"))
tw.append((0,0),"我爱PyMuPDF!",fontsize=36,font=font)
(Rect(-36.891998291015625, -66.1240234375, 274.1679992675781, 38.76397705078125), Point(246.02398681640625, 0.0))
# end point is Point(246.02398681640625, 0.0)
```
The TW rectangle includes a few reserves, which makes the use of its width dubious.
TextWriter internally uses `Font` (mentioned in documentation).

You can be even shorter and create/distroy the `Font` object in one instruction: `fitz.Font("helv").text_length()`.
The `Font` class has an additional method `.char_lengths()` which gives you the length of each single char and also supports font fallbacks:
```python
font.char_lengths("我爱PyMuPDF!",fontsize=36)
(36.0, 36.0, 24.01199984550476, 18.0, 29.98800015449524, 20.015999794006348, 24.01199984550476, 25.992000102996826, 21.996000051498413, 10.007999897003174)
```
